src/main/sensors/opticalflow.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * This file is part of Betaflight.
+ *
+ * Betaflight is free software. You can redistribute this software
+ * and/or modify this software under the terms of the GNU General
+ * Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#include <stdint.h>
+#include <math.h>
+
+#include "platform.h"
+
+#ifdef USE_OPTICALFLOW
+
+#include "build/build_config.h"
+#include "build/debug.h"
+
+#include "common/maths.h"
+#include "common/time.h"
+#include "common/utils.h"
+#include "common/filter.h"
+
+#include "config/config.h"
+#include "config/feature.h"
+
+#include "fc/runtime_config.h"
+
+#include "scheduler/scheduler.h"
+
+#include "pg/pg.h"
+#include "pg/pg_ids.h"
+
+#include "drivers/time.h"
+#include "drivers/rangefinder/rangefinder.h"
+#include "drivers/rangefinder/rangefinder_lidarmt.h"
+
+#include "io/beeper.h"
+
+#include "sensors/sensors.h"
+#include "sensors/gyro.h"
+#include "sensors/acceleration.h"
+#include "sensors/opticalflow.h"
+
+#define OPTICALFLOW_CALIBRATION_DURATION_MS 30000
+#define RATE_SCALE_RESOLUTION (1000.0f)
+
+// static prototypes
+static void applySensorRotation(opticalflowRates_t * dist, opticalflowRates_t * src);
+static void applyLPF(opticalflowRates_t * flowRates);
+
+PG_REGISTER_WITH_RESET_TEMPLATE(opticalflowConfig_t, opticalflowConfig, PG_OPTICALFLOW_CONFIG, 0);
+
+PG_RESET_TEMPLATE(opticalflowConfig_t, opticalflowConfig,
+    .opticalflow_hardware = OPTICALFLOW_NONE,
+    .rotation = 0,
+    .flow_lpf = 0,
+    .flipY = 0
+);
+
+static opticalflow_t opticalflow;
+float cosRotAngle = 1.0f;
+float sinRotAngle = 0.0f;
+static pt2Filter_t xFlowLpf, yFlowLpf;
+
+// ======================================================================
+// =================== Opticalflow Main Functions =======================
+// ======================================================================
+static bool opticalflowDetect(opticalflowDev_t * dev, uint8_t opticalflowHardwareToUse) {
+    UNUSED(dev);
+
+    opticalflowType_e opticalflowHardware = OPTICALFLOW_NONE;
+    requestedSensors[SENSOR_INDEX_OPTICALFLOW] = opticalflowHardwareToUse;
+
+    switch (opticalflowHardwareToUse) {
+        case OPTICALFLOW_MT:
+#ifdef USE_OPTICALFLOW_MT
+            if (mtOpticalflowDetect(dev, rangefinderConfig()->rangefinder_hardware)) {
```

Comment:
IMO it should be `opticalflowHardwareToUse`. Fail if configuration is wrong.

`AUTO` may be added if OF should match RF (and RF supports it)

Reply:
So `USE_OPTICALFLOW_MT` means 'use hardware of rangefinder`. You are right, it is working fine this way in this PR. 

I'm a bit afraid that with more RF and OF devices, it may be harder to configure (or refactor in future). But we can fix it later

Reply:
Not only the USE_OPTICALFLOW_MT, but you also have to set the opticalflow_hardware to MT.

This saves only setting the MT version twice in the cli.
So instead of cli:
set rangefinder_hardware = MTF01
set opticalflow_hardware = MTF01

We do:
set rangefinder_hardware = MTF01
set opticalflow_hardware = MT

And when the switch case finds the opticalflow hardware set to MT, it uses the rangefinder hardware as an argument for the opticalflow detect

Reply:
See also https://github.com/betaflight/betaflight/pull/14042

Reply:
Honestly, I still don't understand why it would not work if we added other sensors, 

for example, if you wanna add a sensor called "SENO" to the optical flow, you will add it to the switch case and set it in the cli opticalflow_hardware to "SENO." In that way, the case OPTICALFLOW_MT will not be entered at all.

You also must write a custom detection function for this sensor and give whatever arguments you want. 

I need to see how the mtOpticalflowDetect(dev, rangefinderConfig()->rangefinder_hardware) will affect your addition of another sensor, which will be handled using another case and another function.

Regarding the AUTO option, sure, we can use it, but I need to understand your point first to know what I am fixing with the AUTO.

Reply:
a summery: 

The function mtOpticalflowDetect is special only for the mt device and will be called only if the opticalflowHardwareToUse is set to MT.

If we want to add another sensor to the opticalflowHardwareToUse, another function will be implemented for such a sensor.

please let me know if I overlooked something

Reply:
The problem is not directly with flow sensor.
One situation is:
- MT driver is altered so that if supports OpticalFlow without exposing Rangefinder functionality (process range packets, even without being configured as Rangefinder)
- user install both TF02 and MTF01, TF02 for range, MTF01 for flow
Setting can be:
`rangefinder_hardware = TF02`
`opticalflow_hardware = MTF01`

But you `opticalflow_hardware = MT` won't work anymore. So you end up with 'MT, MTF01, MTF01P,..,SENO, SENO_MKI, SENO_MKII,..., SUPER, SUPER_V1, SUPER_GIGA, ... `.

Using `AUTO` to mean 'same hardware as rangefinder, if HW supports it` will work with current implementation.
`opticalflow_hardware = AUTO` can be se by default.
And there is no action required from users that are fine with common configuration, even if flow handling is improved/changed in future.

But it is quite possible that I'm overthinking it ...